### ISSUE TITLE
Clink hangs (freezes) when typing invalid command

### ISSUE THREAD

**Issue by eblis** [chrisant996/clink#325]
When I type an invalid command (for example `qasdfalkhdgf`) in the command prompt and wait more than a few ms (so it doesn't happen immediately, but after I stop typing for a few ms or I type space or backspace) the process hangs for a while. From what I can tell clink hangs for about 10s on every freeze _instance_.

I'm not sure what's causing the problem, it could be my antivirus software as I've seen it hang the command prompt when I type an invalid command and execute it .. cmd prompt waits a long time (~2s) before telling me that the command doesn't actually exist.

I can get clink freeze to happen most often by typing an invalid command, typing space and then typing backspace to remove the space, at that point the clink hangs for about 10 seconds.

The hang happens even for the same invalid command, for example I got clink to freeze/hang with `fd` command and after some more tries (typing `fd ` then deleting the space) I got clink to freeze again, so it doesn't look like caching the results helps :(

Also, at one point I managed to get clink to hang and there was no command written on the prompt, I had just deleted the last character (which I think was marked in red as invalid, but I'm not sure)

Usually after clink hangs I get some "quiet time" where it won't get a freeze for a few tens of seconds, but that's not always the case. I was able to get a hang (which took around 10s), then 3 seconds later I got another hang :(

I have created multiple dumps for you to play with for the same cmd.exe process.
[cmd.pid17096 - cmd 1 - 1.zip] -- deleted.
[cmd.pid17096 - cmd 1 - 2.zip] -- deleted.
These 2 files are from the same hang _instance_, I got 2 different dumps while clink was hung
[cmd.pid17096 - cmd 2.zip] -- deleted.
This hang is from the same process, but after the initial freeze resumed and i got clink to freeze again.
[cmd.pid17096 - cmd 3.zip] -- deleted.
And this is from when I got clink to hang for the 3rd time.

**Comment by chrisant996** [comment]
The hang is occurring because the `exec.path` setting is enabled, and the `autosuggest.enable` setting is enabled, and the last two paths in your system `%PATH%` are UNC paths to network file shares.

Similar hangs will occur outside of Clink, as well, any time an unrecognized program name is attempted to be launched.

Some possible solutions:
- Remove the network paths from the `%PATH%` env var.  They will inherently cause hangs in various scenarios on the computer, and it's generally recommended to not have network paths in the system path, because it leads to exactly this kind of hang in lots of programs and scenarios.
- Turn off the `exec.path` setting so that executable completion doesn't search the system `%PATH%`.  That will affect completion, both when manually triggering completion and also when the autosuggest feature invokes completion to look for possible completions.  It uses a coroutine to stay responsive, but the network paths cause network calls to happen, and coroutines can't compensate for that.
- Turn off the `autosuggest.enable` setting.  But then you'll lose suggestions altogether.
- Trim the `autosuggest.strategy` setting to not contain `completion`.  Then autosuggest won't hit the network -- only manually invoking completion for the command/executable word will hit the network.

Alternatively, I could add another setting that controls whether `exec.path` will ignore UNC paths.  But you'll still get hangs in other programs and scenarios, and also hangs in various other scenarios inside Clink.  That's just going to happen when there are network paths included on the system path.

**Comment by eblis** [comment]
Thanks for the quick response and for spotting the issue.

I didn't pay attention to my %PATH% variable, I wasn't the one to add those network paths .. I can remove them, but I'm not so sure if IT will be happy if I do that, I will need to ask :)

If you no longer need the dump files can you delete them or shall I ?

**Comment by chrisant996** [comment]
> If you no longer need the dump files can you delete them or shall I ?

I've deleted them.

I'm making the combination of the `exec.path` setting running inside autosuggest always skip remote drives (UNC paths and also network drives, etc).  There are still other ways that Clink or custom scripts can hang on network paths, and there's not a way to universally fix that for the whole computer, other than to either remove the network paths or to make sure the network is stable.  But this will at least help with general case responsiveness.

**Comment by eblis** [comment]
I've been busy creating a startup script for clink that can remove all network paths from %PATH% variable.
Turns out my network paths are added every time I connect to the VPN :(

```
@echo off
REM echo "Starting with path: %PATH%"

SetLocal EnableDelayedExpansion
set $line=%path%
REM set $line=%$line:(={%
set $line=%$line:)=}%
set $line=%$line: =#%
set $line=%$line:;= %
for %%a in (%$line%) do echo %%a | find /i "\\" || set $newpath=!$newpath!;%%a
set $newpath=!$newpath:#= !
REM set $newpath=%$newpath:{=(%
set $newpath=%$newpath:}=)%

set NEW_PATH=!$newpath:~1!
endlocal & SET PATH=%NEW_PATH%

REM echo "Ended up with path %PATH%"
```
My only problem with it is that's is very slow :(

**Comment by eblis** [comment]
Turns out I had been wrongfully blaming my shitty antivirus software all this time when it fact it was network paths in %PATH% variable :)

**Comment by chrisant996** [comment]
> I've been busy creating a startup script for clink that can remove all network paths from %PATH% variable. Turns out my network paths are added every time I connect to the VPN :(
> 
> ...
> 
> My only problem with it is that's is very slow :(

The `os.setenv("PATH", path_string)` API affects the current CMD shell.

So you could write it as a Clink Lua script, and it would be very fast.

**Comment by chrisant996** [comment]
The fix will be in v1.3.37 when that is published.

**Comment by chrisant996** [comment]
This enhancement is available in v1.3.37 now.
